### PR TITLE
Enable generation in notebooks that use IPUSeq2SeqTrainer

### PR DESCRIPTION
**Pull request by kundaMwiza**
# What does this PR do?

- Enable support for prediction with generation in notebooks

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: notebooks/summarization.ipynb [hunk 1]
```diff
@@ -588,6 +588,10 @@
     "ipu_config = IPUConfig.from_pretrained(\n",
     "    ipu_config_name,\n",
     "    executable_cache_dir=executable_cache_dir,\n",
+    "    # -1 wildcard, \n",
+    "    # split encoder and decoder layers evenly across IPUs \n",
+    "    # for inference     \n",
+    "    inference_layers_per_ipu=[-1]\n",
```
Comment (jimypbr) [inline notebooks/summarization.ipynb:594]: Does this work? this ought to be `[-1, -1]` because you need at least 2 IPUs for encoder decoder.

Comment (kundaMwiza) [reply notebooks/summarization.ipynb:594]: It works if #encoder layers == #decoder layers, e.g. 12 total layers -> [3,3,3,3]

Comment (kundaMwiza) [reply notebooks/summarization.ipynb:594]: Plus, [-1] and [-1] * ipus per replica are treated the same, 
[int(simply total_layers / ipus_per_replica)] * ipu_per_replica, and the remaining layers if any placed on the last ipu

Comment (kundaMwiza) [reply notebooks/summarization.ipynb:594]: Based on this:

```python
if target_number_of_layers is not None:
        if not isinstance(target_number_of_layers, int):
            target_number_of_layers = len(target_number_of_layers)

        # if ipus_per_replica is 1, then put everything on IPU0, ignoring layers_per_ipu
        if ipus_per_replica == 1:
            return [target_number_of_layers]

        elif ipus_per_replica > 1:
            # default/wildcards - split layers evenly over all ipus
            if layers_per_ipu in ([-1], [-1] * ipus_per_replica):
                quotient, remainder = divmod(target_number_of_layers, ipus_per_replica)
                layers_per_ipu = [quotient] * ipus_per_replica
                if remainder > 0:
                    # add any remainder layers to last wildcard IPU
                    layers_per_ipu[-1] += remainder

            # combination of wildcards and integers
            elif -1 in layers_per_ipu and len(layers_per_ipu) == ipus_per_replica:
                wildcard_idxs = [idx for idx, v in enumerate(layers_per_ipu) if v == -1]
                num_wildcard_ipus = len(wildcard_idxs)
                # wildcard_layers = target_num_layers - num_non_wildcard_layers
                num_wildcard_layers = target_number_of_layers - sum([l for l in layers_per_ipu if l != -1])
                quotient, remainder = divmod(num_wildcard_layers, num_wildcard_ipus)
                for idx in wildcard_idxs:
                    layers_per_ipu[idx] = quotient
                if remainder > 0:
                    # add any remainder layers to last wildcard IPU
                    layers_per_ipu[wildcard_idxs[-1]] += remainder

            elif len(layers_per_ipu) != ipus_per_replica:
                raise IncompatibleIPUConfigError(
                    "layers_per_ipu has non-default value set, but its length does not match ipus_per_replica. "
                    f"layers_per_ipu={layers_per_ipu}, ipus_per_replica={ipus_per_replica}. "
                )
            # no wildcards used
            elif sum(layers_per_ipu) != target_number_of_layers:
                raise IncompatibleIPUConfigError(
                    "layers_per_ipu does not define the correct number of layers for the current model."
                    " The current IPU Config specifies IPU assignments for "
                    f"{sum(layers_per_ipu)} layers but there are {target_number_of_layers} layers "
                    f"in the model. layers_per_ipu={layers_per_ipu}"
```

Comment (jimypbr) [reply notebooks/summarization.ipynb:594]: so the `ipus_per_replica` is 4 already from the ipu_config you are loading?

Comment (kundaMwiza) [reply notebooks/summarization.ipynb:594]: its inherited from `training_ipus_per_replica`, so yes. Can optimise by setting `inference_ipus_per_replica`, but then ned to make sure the model fits for the checkpoint that the user chooses

**Review by jimypbr — APPROVED** [review]

**Pull request merged**